### ISSUE TITLE
packagecloud yum repo metadata can break Chef Server if auto-update is enabled

### ISSUE THREAD

**Issue by jeremiahsnapp** [chef/chef-server#154]
When `chef-server-ctl install` is used to install Chef Server add-on packages it first configures packagecloud yum/apt repository metadata. If the system's package manager is configured to auto-update then it will download and install new packages including Chef Server and add-on packages. Just installing new Chef packages does not complete a proper upgrade since other `ctl reconfigure` commands must be run. This can leave the Chef Server in a broken state if any of the Chef services get restarted. 

The Chef PackageCloud repository metadata should be configured but disabled with respect to the system's package manager auto-updater to prevent this from happening in the first place.

**Comment by dreamnite** [comment]
Suggest adding to the end of the post install script for chef-server-core:
if [ $1 -ge 2 ]; then
 /usr/bin/chef-server-ctl reconfigure

or similar for the addons: 
if [$1 -ge 2]; then 
/usr/bin/$ADDON-ctl reconfigure

$1 in the post install script can be used to determine if the package has been installed or upgraded, see: 
http://docs.fedoraproject.org/en-US/Fedora_Draft_Documentation/0.1/html/RPM_Guide/ch09s04s05.html

This should preserve the ability of yum to update the package without causing issues to chef because of a lack of "reconfigure" recipes being run.

**Comment by jeremiahsnapp** [comment]
Hello @dreamnite,

The actual command would need to be `chef-server-ctl upgrade` for chef-server-core but I don't think this is a reliable solution anyway.

Chef Server Tier and HA topologies require more orchestrated steps during the upgrade process. The primary backend needs to be upgraded first and then appropriate files need to be transferred to the other servers in the Chef Server cluster before running the upgrade command on those servers.

I think it is too risky to have a Chef Server (especially a cluster of Chef Servers) upgrade automatically in the first place. The Chef Server is typically a critical part of the infrastructure and should be treated more carefully with scheduled upgrades.

I believe it would be safer to just disable the chef-stable repository metadata with `enabled=0`. When the user is ready to upgrade they just use `yum install chef-server-core --enablerepo=chef-stable`. The `chef-server-install` command can also be configured to use that option when installing other add-ons.

**Comment by jeremiahsnapp** [comment]
I changed the title of this issue because I believe this problem only affects system's using yum.

The default contents of "/etc/apt/apt.conf.d/50unattended-upgrades" on an Ubuntu system would prevent chef-stable packages from being auto upgraded.

```
// Automatically upgrade packages from these (origin:archive) pairs
Unattended-Upgrade::Allowed-Origins {
        "${distro_id}:${distro_codename}-security";
//      "${distro_id}:${distro_codename}-updates";
//      "${distro_id}:${distro_codename}-proposed";
//      "${distro_id}:${distro_codename}-backports";
};
```

**Comment by vinyar** [comment]
Not sure if it's related to this particular issue..
I saw chef-server package come back as one of the updates during yum update, which definitely scared me. Would love for that to not be default behavior.

**Comment by jeremiahsnapp** [comment]
@vinyar yes, this is exactly what this issue is about. This definitely should not be the default behavior but it currently is. The fix is easy so I'm hoping someone will add this to the appropriate bug backlog to get it fixed in the next release.

Just add `enabled false` to the `yum_repository` resource in the code linked below.

https://github.com/chef/chef-server/blob/12.1.0/omnibus/files/private-chef-cookbooks/private-chef/recipes/add_ons_remote.rb#L25-L39

Then the `chef-server-ctl install` command needs to account for this by changing the current `package` resource linked below into a `yum_package` resource for `rhel` platforms. Then you can use the `yum_package` resource's `options` attribute to pass the `--enablerepo=chef-stable` option.

https://github.com/chef/chef-server/blob/12.1.0/omnibus/files/private-chef-cookbooks/private-chef/recipes/add_ons_remote.rb#L47-L51

**Comment by vinyar** [comment]
Just hit this in prod at a client. 

Chef server, reporting and manage were all inadvertently upgraded as part of monthly patching.
Chef servers went down and didn't come up until `chef-server-ctl upgrade` followed by reconfigure for the server, and each feature - manage & reporting were executed.

**Comment by jeremiahsnapp** [comment]
I just realized that the PR https://github.com/chef/chef-server/pull/374 will only disable the chef-stable yum repo in new Chef Server installs. The PR does nothing to disable the chef-stable yum repo in existing Chef Servers.

We still need @chef/lob to come up with a fix that happens when a user runs `chef-server-ctl upgrade` or `chef-server-ctl reconfigure`.

**Comment by jeremiahsnapp** [comment]
Looks like @marcparadise has a great fix for the remaining piece of this issue in this PR.

https://github.com/chef/chef-server/pull/500

**Comment by marcparadise** [comment]
#500 landed in master, so I'm marking this as closed.  This fix will be in the upcoming 12.2 release.
